pkg/azure/ocpgwdeployer.go

```diff
@@ -386,32 +373,81 @@ func (d *ocpGatewayDeployer) deleteGateway() error {
 	ctx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
 	defer cancel()
 
-	for i := 0; i < len(gwNodesList); i++ {
-		if !strings.Contains(gwNodesList[i].Name, submarinerGatewayGW) {
-			err = d.K8sClient.RemoveGWLabelFromWorkerNode(&gwNodesList[i])
-			if err != nil {
-				return errors.Wrapf(err, "failed to remove labels from worker node")
-			}
+	for i := range machineSetList {
+		status.Start(fmt.Sprintf("Deleting the gateway instance %q", machineSetList[i].GetName()))
 
-			publicIPName := gwNodesList[i].Name + "-pub"
+		err = d.msDeployer.DeleteByName(machineSetList[i].GetName(), machineSetList[i].GetNamespace())
+		if err != nil {
+			return status.Error(err, "error deleting the gateway instance from node: %q",
+				machineSetList[i].GetName())
+		}
 
-			err = d.deletePublicIP(ctx, pubIPClient, publicIPName)
-			if err != nil {
-				return errors.Wrapf(err, "failed to delete public-ip")
-			}
+		publicIPName := machineSetList[i].GetName() + "-pub"
+
+		err = d.deletePublicIP(ctx, pubIPClient, publicIPName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete public-ip")
+		}
+
+		status.Success("Successfully deleted the instance")
+	}
+
+	// Cleanup nodes that are not dedicated gateway nodes.
+	gwNodesList, err := d.K8sClient.ListGatewayNodes()
+	if err != nil {
+		return status.Error(err, "error listing the Submariner gateway nodes")
+	}
+
+	gwNodes := gwNodesList.Items
+
+	for i := range gwNodes {
+		if gwNodes[i].GetLabels()[ocp.MachineSetOwnerLabel] == "submariner" {
+			continue
+		}
+
+		// TODO this exists for backward compatibility for version which does not use
+		// "submariner.io/machine-set-owner" label for dedicated g/w node. Remove this
+		// when support ends for Submariner version 14.0 and below.
+		if strings.Contains(gwNodes[i].Name, submarinerGatewayGW) {
+			status.Success("Removing machineset", gwNodes[i].Name[:strings.LastIndex(gwNodes[i].Name, "-")])
+			err = d.msDeployer.DeleteByName(gwNodes[i].Name[:strings.LastIndex(gwNodes[i].Name, "-")],
+				"openshift-machine-api")
 		} else {
-			machineSetName := gwNodesList[i].Name[:strings.LastIndex(gwNodesList[i].Name, "-")]
-			prefix := machineSetName[:strings.LastIndex(gwNodesList[i].Name, "-")]
-			zone := machineSetName[strings.LastIndex(gwNodesList[i].Name, "-")-1:]
+			err = d.K8sClient.RemoveGWLabelFromWorkerNode(&gwNodes[i])
+		}
 
-			machineSet, err := d.initMachineSet(prefix, zone)
-			if err != nil {
-				return err
-			}
+		if err != nil {
+			return errors.Wrapf(err, "failed to cleanup node %q", gwNodes[i].Name)
```

Comment:
We should be consistent returning `status.Error` vs `errors.Wrapf.`

```suggestion
			return status.Error(err, "failed to cleanup node %q", gwNodes[i].Name)
```